src/portal_visualization/client.py

```diff
@@ -0,0 +1,610 @@
+from collections import namedtuple
+import traceback
+from dataclasses import dataclass
+
+# Flask is safe to import since hubmap_commons is a dependency
+from flask import abort, current_app
+import requests
+import json
+from werkzeug.exceptions import HTTPException
+
+
+from .utils import files_from_response
+from .builder_factory import get_view_config_builder
+from .epic_factory import get_epic_builder
+from .builders.base_builders import ConfCells
+
+Entity = namedtuple("Entity", ["uuid", "type", "name"], defaults=["TODO: name"])
+
+
+@dataclass
+class VitessceConfLiftedUUID:
+    vitessce_conf: dict
+    vis_lifted_uuid: str
+
+
+@dataclass
+class PublicationJSONLiftedUUID:
+    publication_json: dict
+    vis_lifted_uuid: str
+
+
+def _get_hits(response_json):
+    """
+    The repeated key makes error messages ambiguous.
+    Split it into separate calls so we can tell which fails.
+    """
+    outer_hits = response_json["hits"]
+    inner_hits = outer_hits["hits"]
+    return inner_hits
+
+
+def _handle_request(url, headers=None, body_json=None):
+    try:
+        response = (
+            requests.post(url, headers=headers, json=body_json)
+            if body_json
+            else requests.get(url, headers=headers)
+        )
+    except requests.exceptions.ConnectTimeout as error:  # pragma: no cover
+        current_app.logger.error(error)
+        abort(504)
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as error:  # pragma: no cover
+        current_app.logger.error(error.response.text)
+        status = error.response.status_code
+        if status in [400, 404]:
+            # The same 404 page will be returned,
+            # whether it's a missing route in portal-ui,
+            # or a missing entity in the API.
+            abort(status)
+        if status in [401]:
+            # I believe we have 401 errors when the globus credentials
+            # have expired, but are still in the flask session.
+            abort(status)
+        raise
+    return response
+
+
+class ApiClient:
+    def __init__(
+        self,
+        groups_token=None,
+        elasticsearch_endpoint=None,
+        portal_index_path=None,
+        ubkg_endpoint=None,
+        assets_endpoint=None,
+        soft_assay_endpoint=None,
+        soft_assay_endpoint_path=None,
+        entity_api_endpoint=None,
+    ):
+        self.groups_token = groups_token
+        self.ubkg_endpoint = ubkg_endpoint
+        self.assets_endpoint = assets_endpoint
+        self.entity_api_endpoint = entity_api_endpoint
+
+        self._elasticsearch_endpoint = elasticsearch_endpoint
+        self._portal_index_path = portal_index_path
+
+        self._soft_assay_endpoint = soft_assay_endpoint
+        self._soft_assay_endpoint_path = soft_assay_endpoint_path
+
+        self.elasticsearch_url = f"{elasticsearch_endpoint}{portal_index_path}"
+        self.soft_assay_url = f"{soft_assay_endpoint}/{soft_assay_endpoint_path}"
```

Comment:
Everything here mostly the same as the `client.py` in `portal-ui` outside of passing the configuration values to the constructor?

Reply:
Yes, no changes to functionality, just switched the environment variables to instance variables.